// quill/modules/table.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Delta = require('../node_modules/quill-delta/dist/Delta.js');
var quill = require('../core/quill.js');
var module$1 = require('../core/module.js');
var table = require('../formats/table.js');

class Table extends module$1.default {
  static register() {
    quill.default.register(table.TableCell);
    quill.default.register(table.TableRow);
    quill.default.register(table.TableBody);
    quill.default.register(table.TableContainer);
  }
  constructor() {
    super(...arguments);
    this.listenBalanceCells();
  }
  balanceTables() {
    this.quill.scroll.descendants(table.TableContainer).forEach(table => {
      table.balanceCells();
    });
  }
  deleteColumn() {
    const [table,, cell] = this.getTable();
    if (cell == null) return;
    // @ts-expect-error
    table.deleteColumn(cell.cellOffset());
    this.quill.update(quill.default.sources.USER);
  }
  deleteRow() {
    const [, row] = this.getTable();
    if (row == null) return;
    row.remove();
    this.quill.update(quill.default.sources.USER);
  }
  deleteTable() {
    const [table] = this.getTable();
    if (table == null) return;
    // @ts-expect-error
    const offset = table.offset();
    // @ts-expect-error
    table.remove();
    this.quill.update(quill.default.sources.USER);
    this.quill.setSelection(offset, quill.default.sources.SILENT);
  }
  getTable() {
    let range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.quill.getSelection();
    if (range == null) return [null, null, null, -1];
    const [cell, offset] = this.quill.getLine(range.index);
    if (cell == null || cell.statics.blotName !== table.TableCell.blotName) {
      return [null, null, null, -1];
    }
    const row = cell.parent;
    const table$1 = row.parent.parent;
    // @ts-expect-error
    return [table$1, row, cell, offset];
  }
  insertColumn(offset) {
    const range = this.quill.getSelection();
    if (!range) return;
    const [table, row, cell] = this.getTable(range);
    if (cell == null) return;
    const column = cell.cellOffset();
    table.insertColumn(column + offset);
    this.quill.update(quill.default.sources.USER);
    let shift = row.rowOffset();
    if (offset === 0) {
      shift += 1;
    }
    this.quill.setSelection(range.index + shift, range.length, quill.default.sources.SILENT);
  }
  insertColumnLeft() {
    this.insertColumn(0);
  }
  insertColumnRight() {
    this.insertColumn(1);
  }
  insertRow(offset) {
    const range = this.quill.getSelection();
    if (!range) return;
    const [table, row, cell] = this.getTable(range);
    if (cell == null) return;
    const index = row.rowOffset();
    table.insertRow(index + offset);
    this.quill.update(quill.default.sources.USER);
    if (offset > 0) {
      this.quill.setSelection(range, quill.default.sources.SILENT);
    } else {
      this.quill.setSelection(range.index + row.children.length, range.length, quill.default.sources.SILENT);
    }
  }
  insertRowAbove() {
    this.insertRow(0);
  }
  insertRowBelow() {
    this.insertRow(1);
  }
  insertTable(rows, columns) {
    const range = this.quill.getSelection();
    if (range == null) return;
    const delta = new Array(rows).fill(0).reduce(memo => {
      const text = new Array(columns).fill('\n').join('');
      return memo.insert(text, {
        table: table.tableId()
      });
    }, new Delta.default().retain(range.index));
    this.quill.updateContents(delta, quill.default.sources.USER);
    this.quill.setSelection(range.index, quill.default.sources.SILENT);
    this.balanceTables();
  }
  listenBalanceCells() {
    this.quill.on(quill.default.events.SCROLL_OPTIMIZE, mutations => {
      mutations.some(mutation => {
        if (['TD', 'TR', 'TBODY', 'TABLE'].includes(mutation.target.tagName)) {
          this.quill.once(quill.default.events.TEXT_CHANGE, (delta, old, source) => {
            if (source !== quill.default.sources.USER) return;
            this.balanceTables();
          });
          return true;
        }
        return false;
      });
    });
  }
}

exports.default = Table;
