// lodash-es/assign.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _assignValue = require('./_assignValue.js');
var _copyObject = require('./_copyObject.js');
var _createAssigner = require('./_createAssigner.js');
var isArrayLike = require('./isArrayLike.js');
var _isPrototype = require('./_isPrototype.js');
var keys = require('./keys.js');

/** Used for built-in method references. */
var objectProto = Object.prototype;

/** Used to check objects for own properties. */
var hasOwnProperty = objectProto.hasOwnProperty;

/**
 * Assigns own enumerable string keyed properties of source objects to the
 * destination object. Source objects are applied from left to right.
 * Subsequent sources overwrite property assignments of previous sources.
 *
 * **Note:** This method mutates `object` and is loosely based on
 * [`Object.assign`](https://mdn.io/Object/assign).
 *
 * @static
 * @memberOf _
 * @since 0.10.0
 * @category Object
 * @param {Object} object The destination object.
 * @param {...Object} [sources] The source objects.
 * @returns {Object} Returns `object`.
 * @see _.assignIn
 * @example
 *
 * function Foo() {
 *   this.a = 1;
 * }
 *
 * function Bar() {
 *   this.c = 3;
 * }
 *
 * Foo.prototype.b = 2;
 * Bar.prototype.d = 4;
 *
 * _.assign({ 'a': 0 }, new Foo, new Bar);
 * // => { 'a': 1, 'c': 3 }
 */
var assign = _createAssigner.default(function(object, source) {
  if (_isPrototype.default(source) || isArrayLike.default(source)) {
    _copyObject.default(source, keys.default(source), object);
    return;
  }
  for (var key in source) {
    if (hasOwnProperty.call(source, key)) {
      _assignValue.default(object, key, source[key]);
    }
  }
});

exports.default = assign;
