// quill/themes/bubble.js
import '../../lodash-es/_baseGetTag.js';
import '../../lodash-es/_baseToString.js';
import '../../lodash-es/_metaMap.js';
import '../../lodash-es/_root.js';
import '../../lodash-es/_LazyWrapper.js';
import '../../lodash-es/wrapperLodash.js';
import '../../lodash-es/_defineProperty.js';
import '../../lodash-es/assign.js';
import '../../lodash-es/assignIn.js';
import '../../lodash-es/assignInWith.js';
import '../../lodash-es/assignWith.js';
import '../../lodash-es/at.js';
import '../../lodash-es/attempt.js';
import '../../lodash-es/bind.js';
import '../../lodash-es/bindAll.js';
import '../../lodash-es/bindKey.js';
import '../../lodash-es/_createRound.js';
import '../../lodash-es/_baseClone.js';
import '../../lodash-es/_LodashWrapper.js';
import '../../lodash-es/_isFlattenable.js';
import '../../lodash-es/_Stack.js';
import '../../lodash-es/_SetCache.js';
import '../../lodash-es/_equalByTag.js';
import '../../lodash-es/isArguments.js';
import '../../lodash-es/isBuffer.js';
import '../../lodash-es/isTypedArray.js';
import '../../lodash-es/_getTag.js';
import '../../lodash-es/_stringToPath.js';
import '../../lodash-es/curry.js';
import '../../lodash-es/curryRight.js';
import '../../lodash-es/defaults.js';
import '../../lodash-es/defaultsDeep.js';
import '../../lodash-es/defer.js';
import '../../lodash-es/delay.js';
import '../../lodash-es/difference.js';
import '../../lodash-es/differenceBy.js';
import '../../lodash-es/differenceWith.js';
import '../../lodash-es/flow.js';
import '../../lodash-es/flowRight.js';
import '../../lodash-es/intersection.js';
import '../../lodash-es/intersectionBy.js';
import '../../lodash-es/intersectionWith.js';
import '../../lodash-es/invoke.js';
import '../../lodash-es/invokeMap.js';
import '../../lodash-es/isArrayBuffer.js';
import '../../lodash-es/isDate.js';
import '../../lodash-es/isPlainObject.js';
import '../../lodash-es/isFinite.js';
import '../../lodash-es/isMap.js';
import '../../lodash-es/_baseIsNative.js';
import '../../lodash-es/_coreJsData.js';
import '../../lodash-es/isRegExp.js';
import '../../lodash-es/isSet.js';
import '../../lodash-es/memoize.js';
import merge from '../../lodash-es/merge.js';
import '../../lodash-es/mergeWith.js';
import '../../lodash-es/method.js';
import '../../lodash-es/methodOf.js';
import '../../lodash-es/toArray.js';
import '../../lodash-es/omit.js';
import '../../lodash-es/over.js';
import '../../lodash-es/overArgs.js';
import '../../lodash-es/overEvery.js';
import '../../lodash-es/overSome.js';
import '../../lodash-es/parseInt.js';
import '../../lodash-es/partial.js';
import '../../lodash-es/partialRight.js';
import '../../lodash-es/pick.js';
import '../../lodash-es/pull.js';
import '../../lodash-es/pullAt.js';
import '../../lodash-es/rearg.js';
import '../../lodash-es/sortBy.js';
import '../../lodash-es/union.js';
import '../../lodash-es/unionBy.js';
import '../../lodash-es/unionWith.js';
import '../../lodash-es/_createSet.js';
import '../../lodash-es/without.js';
import '../../lodash-es/wrapperAt.js';
import '../../lodash-es/xor.js';
import '../../lodash-es/xorBy.js';
import '../../lodash-es/xorWith.js';
import '../../lodash-es/zip.js';
import '../../lodash-es/zipWith.js';
import '../../lodash-es/lodash.default.js';
import Emitter from '../core/emitter.js';
import BaseTheme, { BaseTooltip } from './base.js';
import { Range } from '../core/selection.js';
import Icons from '../ui/icons.js';

const TOOLBAR_CONFIG = [['bold', 'italic', 'link'], [{
  header: 1
}, {
  header: 2
}, 'blockquote']];
class BubbleTooltip extends BaseTooltip {
  static TEMPLATE = ['<span class="ql-tooltip-arrow"></span>', '<div class="ql-tooltip-editor">', '<input type="text" data-formula="e=mc^2" data-link="https://quilljs.com" data-video="Embed URL">', '<a class="ql-close"></a>', '</div>'].join('');
  constructor(quill, bounds) {
    super(quill, bounds);
    this.quill.on(Emitter.events.EDITOR_CHANGE, (type, range, oldRange, source) => {
      if (type !== Emitter.events.SELECTION_CHANGE) return;
      if (range != null && range.length > 0 && source === Emitter.sources.USER) {
        this.show();
        // Lock our width so we will expand beyond our offsetParent boundaries
        this.root.style.left = '0px';
        this.root.style.width = '';
        this.root.style.width = `${this.root.offsetWidth}px`;
        const lines = this.quill.getLines(range.index, range.length);
        if (lines.length === 1) {
          const bounds = this.quill.getBounds(range);
          if (bounds != null) {
            this.position(bounds);
          }
        } else {
          const lastLine = lines[lines.length - 1];
          const index = this.quill.getIndex(lastLine);
          const length = Math.min(lastLine.length() - 1, range.index + range.length - index);
          const indexBounds = this.quill.getBounds(new Range(index, length));
          if (indexBounds != null) {
            this.position(indexBounds);
          }
        }
      } else if (document.activeElement !== this.textbox && this.quill.hasFocus()) {
        this.hide();
      }
    });
  }
  listen() {
    super.listen();
    // @ts-expect-error Fix me later
    this.root.querySelector('.ql-close').addEventListener('click', () => {
      this.root.classList.remove('ql-editing');
    });
    this.quill.on(Emitter.events.SCROLL_OPTIMIZE, () => {
      // Let selection be restored by toolbar handlers before repositioning
      setTimeout(() => {
        if (this.root.classList.contains('ql-hidden')) return;
        const range = this.quill.getSelection();
        if (range != null) {
          const bounds = this.quill.getBounds(range);
          if (bounds != null) {
            this.position(bounds);
          }
        }
      }, 1);
    });
  }
  cancel() {
    this.show();
  }
  position(reference) {
    const shift = super.position(reference);
    const arrow = this.root.querySelector('.ql-tooltip-arrow');
    // @ts-expect-error
    arrow.style.marginLeft = '';
    if (shift !== 0) {
      // @ts-expect-error
      arrow.style.marginLeft = `${-1 * shift - arrow.offsetWidth / 2}px`;
    }
    return shift;
  }
}
class BubbleTheme extends BaseTheme {
  constructor(quill, options) {
    if (options.modules.toolbar != null && options.modules.toolbar.container == null) {
      options.modules.toolbar.container = TOOLBAR_CONFIG;
    }
    super(quill, options);
    this.quill.container.classList.add('ql-bubble');
  }
  extendToolbar(toolbar) {
    // @ts-expect-error
    this.tooltip = new BubbleTooltip(this.quill, this.options.bounds);
    if (toolbar.container != null) {
      this.tooltip.root.appendChild(toolbar.container);
      this.buildButtons(toolbar.container.querySelectorAll('button'), Icons);
      this.buildPickers(toolbar.container.querySelectorAll('select'), Icons);
    }
  }
}
BubbleTheme.DEFAULTS = merge({}, BaseTheme.DEFAULTS, {
  modules: {
    toolbar: {
      handlers: {
        link(value) {
          if (!value) {
            this.quill.format('link', false);
          } else {
            this.quill.theme.tooltip.edit();
          }
        }
      }
    }
  }
});

export { BubbleTooltip, BubbleTheme as default };
