// quill/themes/snow.js
import '../../lodash-es/_baseGetTag.js';
import '../../lodash-es/_baseToString.js';
import '../../lodash-es/_metaMap.js';
import '../../lodash-es/_root.js';
import '../../lodash-es/_LazyWrapper.js';
import '../../lodash-es/wrapperLodash.js';
import '../../lodash-es/_defineProperty.js';
import '../../lodash-es/assign.js';
import '../../lodash-es/assignIn.js';
import '../../lodash-es/assignInWith.js';
import '../../lodash-es/assignWith.js';
import '../../lodash-es/at.js';
import '../../lodash-es/attempt.js';
import '../../lodash-es/bind.js';
import '../../lodash-es/bindAll.js';
import '../../lodash-es/bindKey.js';
import '../../lodash-es/_createRound.js';
import '../../lodash-es/_baseClone.js';
import '../../lodash-es/_LodashWrapper.js';
import '../../lodash-es/_isFlattenable.js';
import '../../lodash-es/_Stack.js';
import '../../lodash-es/_SetCache.js';
import '../../lodash-es/_equalByTag.js';
import '../../lodash-es/isArguments.js';
import '../../lodash-es/isBuffer.js';
import '../../lodash-es/isTypedArray.js';
import '../../lodash-es/_getTag.js';
import '../../lodash-es/_stringToPath.js';
import '../../lodash-es/curry.js';
import '../../lodash-es/curryRight.js';
import '../../lodash-es/defaults.js';
import '../../lodash-es/defaultsDeep.js';
import '../../lodash-es/defer.js';
import '../../lodash-es/delay.js';
import '../../lodash-es/difference.js';
import '../../lodash-es/differenceBy.js';
import '../../lodash-es/differenceWith.js';
import '../../lodash-es/flow.js';
import '../../lodash-es/flowRight.js';
import '../../lodash-es/intersection.js';
import '../../lodash-es/intersectionBy.js';
import '../../lodash-es/intersectionWith.js';
import '../../lodash-es/invoke.js';
import '../../lodash-es/invokeMap.js';
import '../../lodash-es/isArrayBuffer.js';
import '../../lodash-es/isDate.js';
import '../../lodash-es/isPlainObject.js';
import '../../lodash-es/isFinite.js';
import '../../lodash-es/isMap.js';
import '../../lodash-es/_baseIsNative.js';
import '../../lodash-es/_coreJsData.js';
import '../../lodash-es/isRegExp.js';
import '../../lodash-es/isSet.js';
import '../../lodash-es/memoize.js';
import merge from '../../lodash-es/merge.js';
import '../../lodash-es/mergeWith.js';
import '../../lodash-es/method.js';
import '../../lodash-es/methodOf.js';
import '../../lodash-es/toArray.js';
import '../../lodash-es/omit.js';
import '../../lodash-es/over.js';
import '../../lodash-es/overArgs.js';
import '../../lodash-es/overEvery.js';
import '../../lodash-es/overSome.js';
import '../../lodash-es/parseInt.js';
import '../../lodash-es/partial.js';
import '../../lodash-es/partialRight.js';
import '../../lodash-es/pick.js';
import '../../lodash-es/pull.js';
import '../../lodash-es/pullAt.js';
import '../../lodash-es/rearg.js';
import '../../lodash-es/sortBy.js';
import '../../lodash-es/union.js';
import '../../lodash-es/unionBy.js';
import '../../lodash-es/unionWith.js';
import '../../lodash-es/_createSet.js';
import '../../lodash-es/without.js';
import '../../lodash-es/wrapperAt.js';
import '../../lodash-es/xor.js';
import '../../lodash-es/xorBy.js';
import '../../lodash-es/xorWith.js';
import '../../lodash-es/zip.js';
import '../../lodash-es/zipWith.js';
import '../../lodash-es/lodash.default.js';
import Emitter from '../core/emitter.js';
import BaseTheme, { BaseTooltip } from './base.js';
import Link from '../formats/link.js';
import { Range } from '../core/selection.js';
import Icons from '../ui/icons.js';

const TOOLBAR_CONFIG = [[{
  header: ['1', '2', '3', false]
}], ['bold', 'italic', 'underline', 'link'], [{
  list: 'ordered'
}, {
  list: 'bullet'
}], ['clean']];
class SnowTooltip extends BaseTooltip {
  static TEMPLATE = ['<a class="ql-preview" rel="noopener noreferrer" target="_blank" href="about:blank"></a>', '<input type="text" data-formula="e=mc^2" data-link="https://quilljs.com" data-video="Embed URL">', '<a class="ql-action"></a>', '<a class="ql-remove"></a>'].join('');
  preview = this.root.querySelector('a.ql-preview');
  listen() {
    super.listen();
    // @ts-expect-error Fix me later
    this.root.querySelector('a.ql-action').addEventListener('click', event => {
      if (this.root.classList.contains('ql-editing')) {
        this.save();
      } else {
        // @ts-expect-error Fix me later
        this.edit('link', this.preview.textContent);
      }
      event.preventDefault();
    });
    // @ts-expect-error Fix me later
    this.root.querySelector('a.ql-remove').addEventListener('click', event => {
      if (this.linkRange != null) {
        const range = this.linkRange;
        this.restoreFocus();
        this.quill.formatText(range, 'link', false, Emitter.sources.USER);
        delete this.linkRange;
      }
      event.preventDefault();
      this.hide();
    });
    this.quill.on(Emitter.events.SELECTION_CHANGE, (range, oldRange, source) => {
      if (range == null) return;
      if (range.length === 0 && source === Emitter.sources.USER) {
        const [link, offset] = this.quill.scroll.descendant(Link, range.index);
        if (link != null) {
          this.linkRange = new Range(range.index - offset, link.length());
          const preview = Link.formats(link.domNode);
          // @ts-expect-error Fix me later
          this.preview.textContent = preview;
          // @ts-expect-error Fix me later
          this.preview.setAttribute('href', preview);
          this.show();
          const bounds = this.quill.getBounds(this.linkRange);
          if (bounds != null) {
            this.position(bounds);
          }
          return;
        }
      } else {
        delete this.linkRange;
      }
      this.hide();
    });
  }
  show() {
    super.show();
    this.root.removeAttribute('data-mode');
  }
}
class SnowTheme extends BaseTheme {
  constructor(quill, options) {
    if (options.modules.toolbar != null && options.modules.toolbar.container == null) {
      options.modules.toolbar.container = TOOLBAR_CONFIG;
    }
    super(quill, options);
    this.quill.container.classList.add('ql-snow');
  }
  extendToolbar(toolbar) {
    if (toolbar.container != null) {
      toolbar.container.classList.add('ql-snow');
      this.buildButtons(toolbar.container.querySelectorAll('button'), Icons);
      this.buildPickers(toolbar.container.querySelectorAll('select'), Icons);
      // @ts-expect-error
      this.tooltip = new SnowTooltip(this.quill, this.options.bounds);
      if (toolbar.container.querySelector('.ql-link')) {
        this.quill.keyboard.addBinding({
          key: 'k',
          shortKey: true
        }, (_range, context) => {
          toolbar.handlers.link.call(toolbar, !context.format.link);
        });
      }
    }
  }
}
SnowTheme.DEFAULTS = merge({}, BaseTheme.DEFAULTS, {
  modules: {
    toolbar: {
      handlers: {
        link(value) {
          if (value) {
            const range = this.quill.getSelection();
            if (range == null || range.length === 0) return;
            let preview = this.quill.getText(range);
            if (/^\S+@\S+\.\S+$/.test(preview) && preview.indexOf('mailto:') !== 0) {
              preview = `mailto:${preview}`;
            }
            const {
              tooltip
            } = this.quill.theme;
            tooltip.edit('link', preview);
          } else {
            this.quill.format('link', false);
          }
        }
      }
    }
  }
});

export { SnowTheme as default };
